pennylane/transforms/batch_transform.py

```diff
@@ -0,0 +1,239 @@
+            not.
+
+    **Example**
+
+    A valid batch tape transform is a quantum function that satisfies the following:
+
+    - The first argument must be an input tape.
+
+    - Depending on the structure of this input tape, various quantum operations, functions,
+      and templates may be called.
+
+    - Any internal classical processing should use the ``qml.math`` module to ensure
+      the transform is differentiable.
+
+    - The transform should return a tuple containing:
+
+      * Multiple transformed tapes to be executed on a device.
+      * A classical processing function for post-processing the executed tape results.
+        This processing function should have the signature ``f(list[tensor_like]) → Any``.
+        If ``None``, no classical processing is applied to the results.
+
+    For example:
+
+    .. code-block:: python
+
+        @qml.batch_transform
+        def my_transform(tape, a, b):
+            "Generates two tapes, one with all RX replaced with RY,
+            and the other with all RX replaced with RZ."
+
+            tape1 = qml.tape.JacobianTape()
+            tape2 = qml.tape.JacobianTape()
+
+            # loop through all operations on the input tape
+            for op in tape.operations + tape.measurements:
+                if op.name == "RX":
+                    wires = op.wires
+                    param = op.parameters[0]
+
+                    with tape1:
+                        qml.RY(a * qml.math.abs(param), wires=wires)
+
+                    with tape2:
+                        qml.RZ(b * qml.math.abs(param), wires=wires)
+                else:
+                    for t in [tape1, tape2]:
+                        with t:
+                            qml.apply(op)
+
+            def processing_fn(results):
+                return qml.math.sum(qml.math.stack(results))
+
+            return [tape1, tape2], processing_fn
+
+    We can apply this transform to a quantum tape:
+
+    >>> with qml.tape.JacobianTape() as tape:
+    ...     qml.Hadamard(wires=0)
+    ...     qml.RX(-0.5, wires=0)
+    ...     qml.expval(qml.PauliX(0))
+    >>> tapes, fn = my_transform(tape, 0.65, 2.5)
+    >>> print(tapes[0].draw())
+     0: ──H──RY(0.325)──┤ ⟨X⟩
+    >>> print(tapes[1].draw())
+     0: ──H──RZ(1.25)──┤ ⟨X⟩
+
+    We can execute these tapes manually:
+
+    >>> from pennylane.interfaces.batch import execute
+    >>> dev = qml.device("default.qubit", wires=1)
+    >>> res = execute(tapes, dev, interface="autograd", gradient_fn=qml.gradients.param_shift)
+    >>> print(res)
+    [tensor([0.94765073], requires_grad=True), tensor([0.31532236], requires_grad=True)]
+
+    Applying the processing function, we retrieve the end result of the transform:
+
+    >>> print(fn(res))
+    1.2629730888100839
+
+    Alternatively, we may also transform a QNode directly:
+
+    >>> @my_transform(0.65, 2.5)
+    ... @qml.qnode(dev)
+    ... def circuit(x):

```

Comment:
I understand completely what's happening here but I still don't find this very intuitive. If I write a quantum function called `circuit` (or even a QNode) and run `circuit()`, I just expect it to produce the results of the circuit. Maybe this just comes from my preference to apply the transform like a function rather than use it as a decorator. Definitely we should include examples of both when we write up, e.g., a blog post about batch transforms.

Reply:
👍 Agree completely. We should show both examples in the docs.

I must admit I have a personal preference to the decorator syntax just because it feels more compact 😆

Reply:
Have added another example!

Reply:
:100: